strawberryfields/io.py

```diff
@@ -244,6 +246,145 @@ def is_free_param(param):
     return prog
 
 
+def generate_code(prog, eng=None):
+    """Converts a Strawberry Fields program into valid Strawberry Fields code.
+
+    **Example:**
+
+    .. code-block:: python3
+
+        prog = sf.Program(3)
+        eng = sf.Engine("fock", backend_options={"cutoff_dim": 5})
+
+        with prog.context as q:
+            ops.Sgate(2*np.pi/3) | q[1]
+            ops.BSgate(0.6, 0.1) | (q[2], q[0])
+            ops.MeasureFock() | q
+
+        results = eng.run(prog)
+
+        code_str = sf.io.generate_code(prog, eng=eng)
+
+    This will create the following string:
+
+    .. code-block:: pycon
+
+        >>> print(code_str)
+        import strawberryfields as sf
+        from strawberryfields import ops
+
+        prog = sf.Program(3)
+        eng = sf.Engine("fock", backend_options={"cutoff_dim": 5})
+
+        with prog.context as q:
+            ops.Sgate(2*np.pi/3, 0.0) | q[1]
+            ops.BSgate(0.6, 0.1) | (q[2], q[0])
+            ops.MeasureFock() | (q[0], q[1], q[2])
+
+        results = eng.run(prog)
+
+    Args:
+        prog (Program): the Strawberry Fields program
+        eng (Engine): The Strawberryfields engine. If ``None``, only the Program
+            parts will be in the resulting code-string.
+
+    Returns:
+        str: the Strawberry Fields code, for constructing the program, as a string
+    """
+    code_seq = ["import strawberryfields as sf", "from strawberryfields import ops\n"]
+
+    if prog.type == "tdm":
+        code_seq.append(f"prog = sf.TDMProgram(N={prog.N})")
+    else:
+        code_seq.append(f"prog = sf.Program({prog.num_subsystems})")
+
+    # check if an engine is supplied; if so, format and add backend/target
+    # along with backend options
+    if eng:
+        eng_type = eng.__class__.__name__
+        if eng_type == "RemoteEngine":
+            code_seq.append(f'eng = sf.RemoteEngine("{eng.target}")')
+        else:
+            if "cutoff_dim" in eng.backend_options:
+                formatting_str = (
+                    f'"{eng.backend_name}", backend_options='
+                    + f'{{"cutoff_dim": {eng.backend_options["cutoff_dim"]}}}'
+                )
+                code_seq.append(f"eng = sf.Engine({formatting_str})")
+            else:
+                code_seq.append(f'eng = sf.Engine("{eng.backend_name}")')
+
+    # check if program is of TDM type and format the context as appropriate
+    if prog.type == "tdm":
+        # if the context arrays contain pi-values, factor out multiples of np.pi
```

Comment:
Hmm, I'm still not clear on the motivation of factoring out pi :thinking: Might just be me missing something.

Reply:
It simply allows a program that contains e.g. `Sgate(3*np.pi/2)` to actually return a string code representation that looks the same, instead of just returning a float value instead, like `Sgate(4.71238898038)`.

Reply:
Maybe `factor out` is a bit confusing (although that's what we're doing, when it's possible). Not sure if there's a better name. :thinking: